src/test/java/com/google/devtools/build/lib/vfs/FileSystemTest.java

```diff
@@ -1582,11 +1585,29 @@ public void testCanRenameFileToExistingSymlink() throws Exception {
   }
 
   @Test
-  public void testCanRenameSymlinkToExistingFile() throws Exception {
+  public void testCanRenameFileToExistingSymlinkToNonWritableFile() throws Exception {
+    assumeTrue(testFS.supportsSymbolicLinksNatively(xLink.asFragment()));
+
+    Path nonWritableFile = absolutize("non-writable-file");
+    FileSystemUtils.touchFile(nonWritableFile);
+    nonWritableFile.setWritable(false);
+    Path symlink = absolutize("symlink");
+    createSymbolicLink(symlink, nonWritableFile.asFragment());
+    assertThat(symlink.isWritable()).isFalse();
+    xFile.renameTo(symlink); // succeeds
+    assertThat(xFile.exists()).isFalse();
+    assertThat(symlink.isFile()).isTrue();
```

Comment:
This likely doesn't do what you think it does: `Path.isFile` follows symlinks. (There's also an overload with a `followSymlinks` argument.)